File: clients/rust/src/send_bulk.rs
mod proto;

use proto::chat_package::Message;
use proto::chat_package::chat_service_client::ChatServiceClient;
use std::error::Error;
use std::time::Duration;
use tokio::time::sleep;
use tokio_stream::wrappers::ReceiverStream;

const SERVER_ADDR: &str = "http://[::1]:8080";

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let pid = std::process::id();
    // Connect to the gRPC server
    let mut grp_service = ChatServiceClient::connect(SERVER_ADDR).await?;

    let (tx, rx) = tokio::sync::mpsc::channel(128);

    tokio::spawn(async move {
        for i in 0..100 {
            sleep(Duration::from_secs(1)).await;
            let message = Message {
                content: format!("Hi from PID {:?}. This is test #{:?}", pid, i).to_string(),
            };
            println!("Streaming: {:?}", message);
            tx.send(message).await.unwrap();
        }
    });

    let response = grp_service
        .send_bulk_messages(ReceiverStream::new(rx))
        .await?;

    println!("Response: {:?}", response);
    Ok(())
}

File: clients/go/send_bulk.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	proto "grpc-sample-client/protos"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

func main() {
	var addr = "localhost:8080"
	// Set up a connection to the server.
	conn, err := grpc.NewClient(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		log.Fatalf("did not connect: %v", err)
	}
	defer conn.Close()
	gRPCService := proto.NewChatServiceClient(conn)

	// Contact the server and print out its response.
	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
	defer cancel()

	stream, err := gRPCService.SendBulkMessages(ctx)
	if err != nil {
		log.Fatalf("Error starting stream: %v", err)
	}

	var pid = os.Getpid()
	for i := range 100 {
		var request = proto.Message{Content: fmt.Sprintf("Hi from PID %d. This is test %d", pid, i)}
		log.Printf("Streaming: %v", &request)
		err := stream.Send(&request)
		if err != nil {
			log.Fatalf("Error streaming: %v", err)
		}
		time.Sleep(time.Second) // Sleep for 1 second between sends
	}
	response, err := stream.CloseAndRecv()
	if err != nil {
		log.Fatalf("Error ending stream: %v", err)
	}
	log.Printf("Response: %v", response)
}
